File: 兴趣班入门百练讲义/NQ100/NQ099/solution_short.cpp
// NQ099 最省赛程 - 最优解实现
#include < cstring >
#include < iostream >
#include < queue >
#include < algorithm >
using namespace std;
const int N = 1010; //最大城市数量
const int M = 20200; //最大边数（无向图，需存双向边）
const int MAX_CAPACItY = 107;//最大油箱容量
const int INF = 0x3f3f3f3f;//无穷大，代表不可达状态
int oil_price[N]; //每个城市的单位油价
int expenses[N][MAX_CAPACItY];//expenses[city][fuel]：到达city且剩余fuel油量的最小花费
int head[N];
int ver[M];
int next_[M];//避免与std::next重名
int dist_[M];//道路消耗的油量，避免与std::dist重名
int tot = 0; //边的计数器
void add_edge(int u, int v, int d) {
 ver[tot] = v;
 dist_[tot] = d;
 next_[tot] = head[u];
 head[u] = tot++;
}
struct State {
 int city;
 int fuel;
 int cost;
 State(int c, int f, int co) : city(c), fuel(f), cost(co) {}
 bool operator < (const State& other) const {
  return cost > other.cost;
 }
};
// Dijkstra算法：计算从start到end，油箱容量为cap的最小油钱
int Dijkstra(int cap, int start, int end) {
 memset(expenses, 0x3f, sizeof(expenses));
 priority_queue < State > pq;
 expenses[start][0] = 0;
 pq.emplace(start, 0, 0);
 while (!pq.empty()) {
  auto curr = pq.top();
  pq.pop();
  int curr_city = curr.city;
  int curr_fuel = curr.fuel;
  int curr_cost = curr.cost;
  if (curr_cost > expenses[curr_city][curr_fuel]) {
   continue;
  }
  if (curr_city == end) {
   return curr_cost;
  }
  if (curr_fuel < cap) {
   int new_fuel = curr_fuel+1;
   int new_cost = curr_cost+oil_price[curr_city];
   if (new_cost < expenses[curr_city][new_fuel]) {
    expenses[curr_city][new_fuel] = new_cost;
    pq.emplace(curr_city, new_fuel, new_cost);
   }
  }
  for (int i = head[curr_city]; i != -1; i = next_[i]) {
   int next_city = ver[i];
   int fuel_needed = dist_[i];
   // 检查当前油量是否足够到达相邻城市
   if (curr_fuel >= fuel_needed) {
    int new_fuel = curr_fuel-fuel_needed;
    if (curr_cost < expenses[next_city][new_fuel]) {
     expenses[next_city][new_fuel] = curr_cost;
     pq.emplace(next_city, new_fuel, curr_cost);
    }
   }
  }
 }
 return-1;
}
int main() {
 ios::sync_with_stdio(false);
 cin.tie(nullptr);
 int n, m;
 cin > > n > > m;
 for (int i = 0; i < n;++i) {
  cin > > oil_price[i];
 }
 memset(head,-1, sizeof(head));
 tot = 0;
 for (int i = 0; i < m;++i) {
  int u, v, d;
  cin > > u > > v > > d;
  add_edge(u, v, d);
  add_edge(v, u, d);
 }
 int q;
 cin > > q;
 while (q--) {
  int C, S, E;
  cin > > C > > S > > E;
  int res = Dijkstra(C, S, E);
  if (res == -1) {
   cout < < "impossible\n";
  } else {
   cout < < res < < "\n";
  }
 }
 return 0;
}

File: 兴趣班入门百练讲义/NQ100/NQ099/solution_short.py
import sys
import heapq
class Node:
    """优先队列中的节点"""
    def __init__(self, city, fuel, money):
        self.city = city
        self.fuel = fuel
        self.money = money
    def __lt__(self, other):
        """定义小于运算符，使优先队列成为最小堆"""
        return self.money < other.money
def main():
    input_lines = [line.strip() for line in sys.stdin if line.strip()]
    ptr = 0
    n, m = map(int, input_lines[ptr].split())
    ptr += 1
    oil_price = list(map(int, input_lines[ptr].split()))
    ptr += 1
    # 构建邻接表
    adj = [[] for _ in range(n)]
    for _ in range(m):
        u, v, d = map(int, input_lines[ptr].split())
        ptr += 1
        adj[u].append((v, d))
        adj[v].append((u, d))
    q = int(input_lines[ptr])
    ptr += 1
    for _ in range(q):
        c, s, e = map(int, input_lines[ptr].split())
        ptr += 1
        # 初始化费用数组，expenses[city][fuel]表示到达city时剩余fuel的最小花费
        INF = float('inf')
        expenses = [[INF] * (c + 1) for _ in range(n)]
        visited = [[False] * (c + 1) for _ in range(n)]
        # 优先队列
        heap = []
        # 起点状态：城市s，油量0，花费0
        expenses[s][0] = 0
        heapq.heappush(heap, Node(s, 0, 0))
        found = False
        while heap:
            current = heapq.heappop(heap)
            city = current.city
            fuel = current.fuel
            money = current.money
            if visited[city][fuel]:
                continue
            visited[city][fuel] = true
            # 如果到达终点
            if city == e:
                print(money)
                found = true
                break
            # 选项1：在当前城市加1升油
            if fuel < c:
                new_fuel = fuel + 1
                new_money = money + oil_price[city]
                if new_money < expenses[city][new_fuel]:
                    expenses[city][new_fuel] = new_money
                    heapq.heappush(heap, Node(city, new_fuel, new_money))
            # 选项2：前往相邻城市
            for next_city, d in adj[city]:
                if fuel >= d:
                    new_fuel = fuel - d
                    if money < expenses[next_city][new_fuel]:
                        expenses[next_city][new_fuel] = money
                        heapq.heappush(heap, Node(next_city, new_fuel, money))
        if not found:
            print("impossible")
if __name__ == "__main__":
    main()